kvserver: do not report diff in consistency checks

pav-kv opened the pull request:
This commit removes reporting of the diff between replicas in case of consistency check failures. With the increased range sizes the previous approach has become infeasible.

One possible way to inspect an inconsistency after this change is:
1. Run `cockroach debug range-data` tool to extract the range data from each replica's checkpoint.
2. Use standard OS tools like `diff` to analyse them.

In the meantime, we are researching the UX of this alternative approach, and seeing if there can be a better tooling support.

Part of #21128
Epic: none

Release note (sql change): The `crdb_internal.check_consistency` function now does not include the diff between inconsistent replicas, should they occur. If an inconsistency occurs, the storage engine checkpoints should be inspected. This change is made because previously the range size limit has been increased from 64 MiB to O(GiB), so inlining diffs in consistency checks does not scale.

pavelkalinnikov commented:
This probably can be split into 3 commits:
- Delete diff reporting in the consistency check report. This is a short one, but the one that changes behaviour.
- Remove diffing code, as it's used only for testing.
- Remove the "snapshot" code, as it's used only for testing.

tbg reviewed (approved):
> Reviewed 15 of 15 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/89502)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @erikgrinaker and @pavelkalinnikov)

___
*[`pkg/kv/kvserver/replica_consistency.go` line 243 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/89502#-NDl8iU13P16vfThr2yW:-NDl8iU13P16vfThr2yX:b-fu0gxa) ([raw file](https://github.com/cockroachdb/cockroach/blob/ca91e2c2fdce5d60b7f84943b804ac8199d65d90/pkg/kv/kvserver/replica_consistency.go#L243)):*
> ```Go
> 	if args.Checkpoint {
> 		// A diff was already printed. Return because all the code below will do
> 		// is request another consistency check, with a diff and with
> ```

comment needs update
___
*[`pkg/kv/kvserver/replica_consistency.go` line 483 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/89502#-NDl8t6x7qvjXBSEVg3N:-NDl8t6yE8Vuj_IUNkxd:b-d9ei79) ([raw file](https://github.com/cockroachdb/cockroach/blob/ca91e2c2fdce5d60b7f84943b804ac8199d65d90/pkg/kv/kvserver/replica_consistency.go#L483)):*
> ```Go
> 
> // ChecksumRangeForTesting returns a checksum over the KV data of the given
> // range. Only for testing.
> ```

Shouldn't this live in `helpers_test.go` so that it's not in production code?


<!-- Sent from Reviewable.io -->

pav-kv committed:
kvserver: do not report diff in consistency checks

This commit removes reporting of the diff between replicas in case of
consistency check failures. With the increased range sizes the previous
approach has become infeasible.

One possible way to inspect an inconsistency after this change is:
1. Run `cockroach debug range-data` tool to extract the range data from
   each replica's checkpoint.
2. Use standard OS tools like `diff` to analyse them.

In the meantime, we are researching the UX of this alternative approach, and
seeing if there can be a better tooling support.

Release note (sql change): The crdb_internal.check_consistency function now
does not include the diff between inconsistent replicas, should they occur. If
an inconsistency occurs, the storage engine checkpoints should be inspected.
This change is made because previously the range size limit has been increased
from 64 MiB to O(GiB), so inlining diffs in consistency checks does not scale.

pav-kv reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/89502)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @erikgrinaker and @tbg)

___
*[`pkg/kv/kvserver/replica_consistency.go` line 243 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/89502#-NDl8iU13P16vfThr2yW:-NE0oBUrCSr5Cdd_DPk1:b-xdhroi) ([raw file](https://github.com/cockroachdb/cockroach/blob/ca91e2c2fdce5d60b7f84943b804ac8199d65d90/pkg/kv/kvserver/replica_consistency.go#L243)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

comment needs update
</blockquote></details>

Done. Also updated a few other comments.
___
*[`pkg/kv/kvserver/replica_consistency.go` line 483 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/89502#-NDl8t6x7qvjXBSEVg3N:-NE0oK8i0MHhGTzRmghk:b-pbj4wz) ([raw file](https://github.com/cockroachdb/cockroach/blob/ca91e2c2fdce5d60b7f84943b804ac8199d65d90/pkg/kv/kvserver/replica_consistency.go#L483)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

Shouldn't this live in `helpers_test.go` so that it's not in production code?
</blockquote></details>

Good idea. This PR no longer modifies this func, so will address it in the next one.


<!-- Sent from Reviewable.io -->

erikgrinaker reviewed (approved):
LGTM for now. We should consider extending the fatal error message with some instructions on how to go about debugging the inconsistency, once we figure out the tooling.

Why aren't we removing `replica_consistency_diff.go`? We should no longer need it, and if we want to add similar tooling we can dig it out from the Git history.

pavelkalinnikov commented:
I reduced this PR to only remove the diff reporting behaviour (+ proto fields that support it).

pavelkalinnikov commented:
> Why aren't we removing replica_consistency_diff.go? We should no longer need it, and if we want to add similar tooling we can dig it out from the Git history.

I did that first, but then decided to separate it for ease of review. Also, while thinking about the best tooling support, we may need diffs back; so will remove them a bit later if we indeed don't need them. Agree that it could be restored from git, but decided to take smaller steps here (don't think either way is particularly better).

pavelkalinnikov commented:
bors r=erikgrinaker,tbg
